ethpm/contract.py

```diff
@@ -30,46 +30,33 @@ def __init__(self, address: bytes = None, **kwargs: Dict[str, Any]) -> None:
         super(LinkableContract, self).__init__(address=checksummed, **kwargs)
 
     @classmethod
-    def constructor(self, *args: Any, **kwargs: Any) -> bool:
-        if self.has_linkable_bytecode() and not self.is_bytecode_linked:
+    def factory(cls, web3, class_name=None, **kwargs):
+        cls.needs_bytecode_linking = False
+        if "deployment_link_refs" in kwargs and kwargs['deployment_link_refs']:
+            if not is_prelinked_bytecode(to_bytes(hexstr=kwargs['bytecode']), kwargs['deployment_link_refs']):
+                cls.needs_bytecode_linking = True
+        return super(LinkableContract, cls).factory(web3, class_name, **kwargs)
```

Comment:
@pipermerriam can I get another :eyes: on this. Not sure if this is the best way to implement the `needs_bytecode_linking` factory. It's working as intended, but especially w/ line 69-70 feels kind of hacky to me

Reply:
Should look something like this.

```python
class LinkableContract(Contract):
    ...
    # this is what allows this value to be set with `kwargs` in the `factory` method
    needs_bytecode_linking = None

    @classmethod
    def factory(cls, web3, class_name=None, **kwargs):
        needs_bytecode_linking = ...  # whatever check needs to be done to figure this out
        kwargs = assoc(kwargs, 'needs_bytecode_linking', needs_bytecode_linking)
        return super(LinkableContract, cls).factory(web3, class_name, **kwargs)
```

The main `factory` method uses `PropertyCheckingFactory` which is a metaclass which injects all of the given `kwargs` as class properties, but **only if** the class or one of the base classes already has the property (to restrict accidental injection of new class properties due to typos, etc).